File: backend/app/agent/orchestrator.py
from __future__ import annotations

import asyncio
from typing import Any, Callable

from app.agent.classifier import classify_session
from app.agent.curator import curate_window
from app.agent.event_designer import design_events
from app.agent.normalizer import normalize_event
from app.adapters import NotionWriter, capture_and_upload
from app.config import settings
from app.models import EventWindow, SessionState


class StudyLensAgent:
    def __init__(
        self,
        session: SessionState,
        notion_writer: NotionWriter,
        activity_callback: Callable[..., None] | None = None,
    ) -> None:
        self.session = session
        self.notion = notion_writer
        self.emit = activity_callback or (lambda **_: None)
        self._buffer = []
        self._window_start = 0.0
        self._window_id = 0
        self._classified = False
        self._events_designed = False

    def ingest_raw_event(self, raw_event: dict) -> None:
        normalized = normalize_event(raw_event)
        if normalized is None:
            return
        self._buffer.append(normalized)
        self.session.events_received += 1

    def _window_ready(self) -> bool:
        if not self._buffer:
            return False
        if self._window_start == 0.0:
            self._window_start = min(e.timestamp_start for e in self._buffer)
        now = max(e.timestamp_end for e in self._buffer)
        threshold = (
            settings.curator_initial_window_seconds
            if self._window_id == 0
            else settings.curator_window_seconds
        )
        return (now - self._window_start) >= threshold

    def _close_window(self) -> EventWindow | None:
        if not self._window_ready():
            return None
        end_time = max(e.timestamp_end for e in self._buffer)
        window = EventWindow(
            window_id=self._window_id,
            start_time=self._window_start,
            end_time=end_time,
            events=list(self._buffer),
        )
        self._buffer = []
        self._window_start = 0.0
        self._window_id += 1
        return window

    async def process_if_ready(
        self,
        conn: Any | None = None,
        scene_index: Any | None = None,
        ws_connection_id: str = "",
    ) -> None:
        window = self._close_window()
        if window is None:
            return

        decision = await curate_window(window, previous_title=self.session.last_concept_title)
        self.session.windows_processed += 1

        if not decision.should_write:
            self.session.windows_skipped += 1
            self.session.last_skip_reason = decision.skip_reason or "skipped"
            self.emit(
                category="skip",
                icon="⏭️",
                label=f"Skipped window {window.window_id}: {self.session.last_skip_reason}",
                detail=(decision.summary or "")[:180],
            )
            return

        await self.notion.write_decision(decision)
        # Flush first section immediately to reduce visible startup lag in Notion.
        if window.window_id == 0:
            await self.notion.flush()
        self.session.windows_written += 1
        self.session.last_concept_title = decision.concept_title
        self.emit(category="section", icon="✓", label=f"Section: {decision.concept_title}")

        # Always attempt screenshot per written window for demo reliability.
        if self.session.display_rtstream_id:
            asyncio.create_task(self._attach_screenshot(decision))

        if not self._classified:
            result = await classify_session(window)
            content_type = str(result.get("content_type", "lecture"))
            self.session.classifier_result = content_type
            self.emit(category="classification", icon="🧠", label=f"Classified: {content_type}")
            self._classified = True
            if conn is not None and scene_index is not None and ws_connection_id:
                designed = await design_events(conn, scene_index, ws_connection_id, content_type)
                self.session.events_designed = [e.get("label", "") for e in designed if e.get("label")]
                for ev in designed:
                    self.emit(
                        category="tool_call",
                        icon="🔧",
                        label=f"Agent registered: {ev.get('label', 'unknown')}",
                        detail=f"-> {ev.get('product_action', '')}",
                    )
                self._events_designed = True

    async def _attach_screenshot(self, decision) -> None:
        try:
            upload_id = await asyncio.to_thread(
                capture_and_upload,
                self.session.display_rtstream_id,
                decision.window_start,
                decision.window_end,
                decision.visual_evidence.target_time_hint,
            )
            if upload_id:
                await self.notion.append_image_file_upload(
                    upload_id,
                    caption=decision.visual_evidence.reason or "",
                )
                self.session.screenshots_captured += 1
                self.emit(category="screenshot", icon="📷", label="Screenshot attached")
        except Exception:
            self.session.screenshot_failures += 1

    async def flush_on_stop(self) -> None:
        if self._buffer:
            end_time = max(e.timestamp_end for e in self._buffer)
            window = EventWindow(
                window_id=self._window_id,
                start_time=self._window_start or end_time,
                end_time=end_time,
                events=list(self._buffer),
            )
            self._buffer = []
            decision = await curate_window(window, previous_title=self.session.last_concept_title)
            if decision.should_write:
                await self.notion.write_decision(decision)

    async def answer_question(self, question: str) -> str:
        return f"Q&A placeholder. Received: {question[:120]}"

File: backend-go/internal/agent/orchestrator.go
package agent

import (
	"context"
	"fmt"
	"sync"

	"github.com/rs/zerolog/log"
	"github.com/sangeetmore/novo-transcriber/backend-go/internal/adapters"
	"github.com/sangeetmore/novo-transcriber/backend-go/internal/config"
	"github.com/sangeetmore/novo-transcriber/backend-go/internal/models"
)

// NoteItAgent is the core orchestrator.
type NoteItAgent struct {
	session *models.SessionState
	notion  *adapters.NotionWriter
	llm     *adapters.Client
	cfg     *config.Config
	emit    func(category, icon, label, detail string, metadata map[string]any)

	mu             sync.Mutex
	buffer         []models.NormalizedEvent
	windowStart    float64
	windowID       int
	classified     bool
	eventsDesigned bool
}

func NewNoteItAgent(
	session *models.SessionState,
	notion *adapters.NotionWriter,
	llm *adapters.Client,
	cfg *config.Config,
	emit func(category, icon, label, detail string, metadata map[string]any),
) *NoteItAgent {
	return &NoteItAgent{
		session:     session,
		notion:      notion,
		llm:         llm,
		cfg:         cfg,
		emit:        emit,
		windowStart: 0.0,
		windowID:    0,
	}
}

func (a *NoteItAgent) IngestRawEvent(raw map[string]any) {
	evt := NormalizeEvent(raw)
	if evt == nil {
		return
	}

	a.mu.Lock()
	a.buffer = append(a.buffer, *evt)
	a.mu.Unlock()

	a.session.IncrEventsReceived()
}

func (a *NoteItAgent) windowReady() bool {
	a.mu.Lock()
	defer a.mu.Unlock()

	if len(a.buffer) == 0 {
		return false
	}

	// Calculate timespan
	minStart := a.buffer[0].TimestampStart
	maxEnd := a.buffer[0].TimestampEnd
	for _, e := range a.buffer {
		if e.TimestampStart < minStart {
			minStart = e.TimestampStart
		}
		if e.TimestampEnd > maxEnd {
			maxEnd = e.TimestampEnd
		}
	}

	span := maxEnd - a.windowStart
	threshold := a.cfg.CuratorInitialWindowSeconds
	if a.windowID > 0 {
		threshold = a.cfg.CuratorWindowSeconds
	}

	return span >= float64(threshold)
}

func (a *NoteItAgent) closeWindow() *models.EventWindow {
	a.mu.Lock()
	defer a.mu.Unlock()

	if len(a.buffer) == 0 {
		return nil
	}

	maxEnd := a.buffer[0].TimestampEnd
	for _, e := range a.buffer {
		if e.TimestampEnd > maxEnd {
			maxEnd = e.TimestampEnd
		}
	}

	w := &models.EventWindow{
		WindowID:  a.windowID,
		StartTime: a.windowStart,
		EndTime:   maxEnd,
		Events:    make([]models.NormalizedEvent, len(a.buffer)),
	}
	copy(w.Events, a.buffer)

	a.windowID++
	a.windowStart = maxEnd
	a.buffer = nil

	return w
}

func (a *NoteItAgent) ProcessIfReady(ctx context.Context) error {
	if !a.windowReady() {
		return nil
	}

	window := a.closeWindow()
	if window == nil {
		return nil
	}

	a.session.RLock()
	prevTitle := a.session.LastConceptTitle
	a.session.RUnlock()

	decision, err := CurateWindow(ctx, *window, prevTitle, a.llm, a.cfg)
	if err != nil {
		log.Warn().Err(err).Msg("curate window failed, using fallback")
		decision = deterministicFallback(*window)
	}

	a.session.IncrWindowsProcessed()

	if !decision.ShouldWrite {
		a.session.IncrWindowsSkipped()
		a.emit("agent", "⏭️", "Skipped segment", decision.SkipReason, nil)
		return nil
	}

	a.notion.WriteDecision(ctx, decision)
	a.session.IncrWindowsWritten()

	a.session.Lock()
	a.session.LastConceptTitle = decision.ConceptTitle
	a.session.Unlock()

	a.emit("agent", "📝", "Notes generated", decision.ConceptTitle, nil)

	if window.WindowID == 0 {
		a.notion.Flush(ctx) // flush immediately for the first window
	}

	// Screenshot capturing logic would go here
	// go a.attachScreenshot(context.Background(), decision)

	return nil
}

func (a *NoteItAgent) FlushOnStop(ctx context.Context) error {
	window := a.closeWindow()
	if window == nil || !window.HasContent() {
		return nil
	}

	a.session.RLock()
	prevTitle := a.session.LastConceptTitle
	a.session.RUnlock()

	decision, _ := CurateWindow(ctx, *window, prevTitle, a.llm, a.cfg)
	if decision.ShouldWrite {
		a.notion.WriteDecision(ctx, decision)
		a.session.IncrWindowsWritten()
		a.emit("agent", "📝", "Final notes generated", decision.ConceptTitle, nil)
	}
	return nil
}

func (a *NoteItAgent) AnswerQuestion(ctx context.Context, question string) (string, error) {
	return fmt.Sprintf("Q&A placeholder. Received: %s", question), nil
}
